### ISSUE TITLE
[no-unnecessary-type-assertion] false positive due to readonly array

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3777]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": [2]
  }
}
```

```TS
export type Readonlyable<T> = Readonly<T> | T

export type AnyArray<T> = Readonlyable<T[]>

export type Arrayable<T, R extends boolean = false> = [R] extends [never]
  ? T | T[]
  : R extends true
  ? Readonly<T> | readonly T[]
  : R extends false
  ? AnyArray<T> | Readonlyable<T>
  : never


export type TranslateKey<T extends string = string> =
  | Arrayable<number | string>
  | Partial<Record<T, string>>


export const isObjectType = <T>(obj: unknown): obj is T => Object.prototype.toString.call(obj) === '[object Object]'

export const arrayify = <T, R = T extends Array<infer S> ? S : T>(
  ...args: T[]
) =>
  args.reduce<Array<Exclude<R, null | undefined>>>((arr, curr) => {
    arr.push(...(Array.isArray(curr) ? curr : curr == null ? [] : [curr]))
    return arr
  }, [])


declare const base: <L extends string>(key: TranslateKey<L>) => string

export const test = <L extends string>(key: TranslateKey<L>) => {
  const ns = ['1']
  base(isObjectType<Record<L, string>>(key) ? key : [...ns, ...(arrayify(key) as Array<string | number>)]) // reported here
}
```

The fixed code emits error by TS: [playground](https://www.typescriptlang.org/play?jsx=0&ts=4.3.5#code/KYDwDg9gTgLgBDAnmYcBKwCGATCA7AG0UwCMDgAeAFQD44BedLXQxaugHzioCgfRIsBMlQBBPIlFQomNrQZMc+IqXLUA2gF0afAdHhIUcKTOJlKVADTo4oGMDzYAznBIQI5THgUAzTASdgOkZ1NE1bEHtHF3U8YAA3YChNHjg4AH5uOC4qLVS4AC4bOwdnBCgAV2B8zIwlVnZsuChmZURuPLSitAiosr8A6rTM8UlpWUauOpYVc3Z8orjEqD5+cH1hIyoZPCcCTHsAaWA5XtKXJxgoAEs8AHMFS5v74PyuE1lVSjwKgFsSJJNJ63O46NJcAAKmFg138FAwAGNoNhqNZgS8dLp1kIkbt4NcnAB5EgAK2ACJgVBECnYAAoIKSihU8ABrPAQADueAAlEUGSS4ASsvQ6MSyRSAHRgKAQGCykQSuUAZSuIIlCP8BHppO5DHojAA5Op+eT4GLTZoDVjBPBcZc4NDTNcfO1GKibIwqGdosZxmxbj5AUq6JklYVuDRafkJTHoXcnEVcildSL8nGnBKWtgKgjKB82ABREAIggVbCUNDWH4EAhNZnlny3YDYGit2m0x3WBEVaQpugAb3yaUdUoqTgAFrSYxLafmJQT87Tu72MnBl1Bw+u9XBq7XMlpw+p15pudyh81gDAe95HfkAL7WLRnvjlkvQ1B2+AkTCBIoUAAy3plOioK0iyJyJjsewHMAxxsP+NB9nAIHWhsn4IMA9puoBJQ+iBkbgYgkFeNBRwnABiEMAO+TobsCjqAaACMBopGk36BLSBLmhSVIoPC5LIgBaKqhiYEnLqmSEYe067NY04dn6zqIGJiCniePB3jwQA)

**Expected Result**

No error

**Actual Result**

Error

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.2` |
| `@typescript-eslint/parser`        | `4.29.2` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.22.5` |

**Comment by JounQin** [comment]
related https://github.com/microsoft/TypeScript/issues/17002

**Comment by bradzacher** [comment]
This is due to the direct cast on a generic function type.

In a nutshell you have this:

```ts
function fn<L extends string>(key: TranslateKey<L>) {
  if (!isObjectType<Record<L, string>>(key)) {
    arrayify(key);
  }
}
```

Which in a nutshell is simplified to this:

```ts
function fn<L extends string>(key: string | number | (string | number)[]) {
  arrayify<
    string | number | (string | number)[],
    string | number | (string | number)[],
  >(key);
  // type === (string | number)[]
}
```

But when you add the assertion, what happens is this:
```ts
function fn<L extends string>(key: string | number | (string | number)[]) {
  arrayify<
    (string | number)[],
    string | number
  >(key) as (string | number)[];
  // return type === (string | number)[]
}
```

What's happening is that TS is using the assertion to "backfill" the generic types.
But what the heck? Why is this happening?

Well it's because you have a generic that's only used in an output location in the function signature!

What do I mean by that? your `arrayify` function boils down to this signature:
```ts
declare function arrayify<T, R>(...args: T[]): Array<R>;
```

Note how `R` is not used in an input location and instead is only used in an output location? That's the problem. This is an ***anti-pattern*** in TypeScript and it can lead to weirdness like you see.

You can inspect all of the types in a tool like [ts-ast-viewer](https://ts-ast-viewer.com/#code/KYDwDg9gTgLgBDAnmYcBKwCGATCA7AG0UwCMDgAeAFQD44BedLXQxaugHzioCgfRIsBMlQBBPIlFQomNrQZMc+IqXLUA2gF0afAdHhIUcKTOJlKVADTo4oGMDzYAznBIQI5THgUAzTASdgOkZ1NE1bEHtHF3U8YAA3YChNHjg4AH5uOC4qLVS4AC4bOwdnBCgAV2B8zIwlVnZsuChmZURuPLSitAiosr8A6rTM8UlpWUauOpYVc3Z8orjEqD5+cH1hIyoZPCcCTHsAaWA5XtKXJxgoAEs8AHMFS5v74PyuE1lVSjwKgFsSJJNJ63O46NJcAAKmFg138FAwAGNoNhqNZgS8dDxsMAEfsWnAfBU8AiYNd8HBrk4APIkABWOJgVBEqJoAAoIHSikSANZ4CAAdzwAEoihzaRSXFQANx8bG46GoQnE0nk6Gma4+OTWHqMKhnaLGcZsW4+QEAZTomTNhW4bPyADpHdC7k4irkUiLDaYKABREC4irY+HWH4EAhNInYny3YDYGg0GVYnF4xVEklk7xq2QaxAAJhZrIdTqgLrdeU9HzYfoDQb1JQNlYoJvNlrg1rdIYqYYjjmA0bicYTsuTCoJaZV3hImECFAAMvqyujQazuSc3Ts9gdgMc2LOaJ6l4mlenyfZLnOFxcriC2avEOuvJujic5-u4ABvfJI3bwXYKdQAOQAIwAZoMppBqcCspSNL0iSTIoPCOLInOaLXhiK4nEKQofvkaRToEmGIEK4FwAAvrYgy4WkaTfpccCYAoWaIDmREkXhcB0fAJBMUarF3jh05ehMS5ND8-xJEOHFcZxvHqpquZsaRtH4PR2BydmClsQxLiNqJXDiQCUBDmkZE8GRQA) and you can see that when you add the assertion - TS ***changes the generics passed into `arrayify`***. This change in generics changes the return type.

So when our rule inspects the types, it just sees that the return type matches the assertion. Which it obviously reports against. There is nothing we can do here - we can only act on the information given to us by TS.

You need to change your types such that all generics are used in an input position.

Here are several issues that stem from the same problem: #391, #528, #1410
